contracts/atomic-order-example/src/contract.rs

```diff
@@ -0,0 +1,188 @@
+        .debug(format!("SC ----> Funds info: {}", info.funds[0]).as_str());
+    let config = STATE.load(deps.storage)?;
+    let contract = env.contract.address;
+    let subaccount_id = default_subaccount_id(&contract);
+    let min_deposit = price * quantity;
+    let message_deposit = FPDecimal::from(info.funds[0].amount.u128());
+    if message_deposit < min_deposit {
+        return Err(ContractError::CustomError {
+            val: format!(
+                "Deposit: {} below min_deposit: {}",
+                message_deposit, min_deposit
+            ),
+        });
+    }
+    let order = SpotOrder::new(
+        price,
+        quantity,
+        true,
+        false,
+        true,
+        &config.market_id,
+        &subaccount_id,
+        contract.as_str(),
+    );
+
+    let coins = &info.funds[0];
+    let deposit_message = SubMsg::new(create_deposit_msg(
+        contract.to_string(),
+        subaccount_id,
+        coins.clone(),
+    ));
+    let order_message = SubMsg::reply_on_success(
+        create_spot_market_order_msg(contract.into_string(), order),
+        ATOMIC_ORDER_REPLY_ID,
+    );
+    let response = Response::new()
+        .add_submessage(deposit_message)
+        .add_submessage(order_message);
+
+    let cache = SwapCacheState { sender_address: info.sender.to_string(), deposited_amount: coins.clone() };
+    CACHE.save(deps.storage, &cache)?;
+
+    return Ok(response);
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn reply(
+    deps: DepsMut,
+    _env: Env,
+    msg: Reply,
+) -> Result<Response<InjectiveMsgWrapper>, StdError> {
+    // deps.api.debug(format!("SC ----> Received reply (data): {}", msg.clone().result.unwrap().data.unwrap()).as_str(), );
+    // let events = msg.clone().result.unwrap().events;
+    // deps.api.debug(format!("SC ----> Received reply (events count): {}", events.len()).as_str());
+    match msg.id {
+        ATOMIC_ORDER_REPLY_ID => handle_atomic_order_reply(deps, msg),
+        id => Err(StdError::generic_err(format!("Unknown reply id: {}", id))),
+    }
+}
+
+fn handle_atomic_order_reply(
+    deps: DepsMut,
+    msg: Reply,
+) -> Result<Response<InjectiveMsgWrapper>, StdError> {
+    let dec_scale_factor: FPDecimal = FPDecimal::from(1000000000000000000 as i128);
+    let mut data = msg.result.unwrap().data.unwrap().to_vec();
+    let _ = parse_protobuf_string(&mut data, 1); // order hash - we need to read to advance pointer
+
+    let trade_result = parse_protobuf_bytes(&mut data, 2).unwrap();
+    let mut trade_data = trade_result.unwrap().to_vec();
+    let field1 = parse_protobuf_string(&mut trade_data, 1).unwrap();
+    let field2 = parse_protobuf_string(&mut trade_data, 2).unwrap();
+    let field3 = parse_protobuf_string(&mut trade_data, 3).unwrap();
+    let quantity = FPDecimal::from_str(&field1)? / dec_scale_factor;
+    let price = FPDecimal::from_str(&field2)? / dec_scale_factor;
+    let fee = FPDecimal::from_str(&field3)? / dec_scale_factor;
```

Comment:
In theory it shouldn't matter, as this shouldn't panic the whole chain, just rollback this SC, but as this is supposed to be educational example it makes sense to avoid unnecessary unwrap - changed